### DISCUSSION TITLE
Directions in Google Maps are starting from strange starting point

### DISCUSSION THREAD

**Discussion by mikepoconnor4** [code4recovery/12-step-meeting-list#857] [Q&A] [answered]
One of our users noted that the directions link from TSML are starting from a strange starting point that has nothing to do with their location. We tried and got the same starting location (a restaurant named "Oh My Grill") - we also simulated being in other countries via VPN and got strange starting points in other places as well. We also tried directions from other integroup sites such as Seattle and SF/Marin and got strange results there also which started from the same "Oh My Grill" starting point.

Here's an example of one of our physical meetings. https://oahuaa.org/oahu-meetings/?type=active&search=%28physical%29&meeting=kukui-hele-po-3

The link for directions on that page is:
https://www.google.com/maps?saddr=Current+Location&daddr=21.2679075%2C-157.8224446&q=2709+Kalakaua+Ave%2C+Honolulu%2C+HI+96815%2C+USA

But that's not what the current Google Maps developer doc says to specify. It says to add a /dir/ in the URL, add an api=1 parameter, no &saddr (start address) parameter (no parameter yields current location), and rather than &daddr (destination address) the parameter is actually &destination.

Google Maps developer doc: https://developers.google.com/maps/documentation/urls/get-started#directions-action

So when I replace the above TSML direction string with the correct parameters. the map gives me directions from my house:
https://www.google.com/maps/dir/?api=1&destination=2709+Kalakaua+Ave%2C+Honolulu%2C+HI+96815%2C+USA

Is the direction URL something that TSML generates or something that we control? We noticed that Seattle's URL has the same pattern as we do, while SF/Marin has a link back to the same meeting page with a "/#" appended. We also tried LA's links - their URL coding is different but their directions pop up with an unspecified starting point that you then need to select.

Thanks in advance for looking into this. You guys are always so responsive and provide a great plugin and service.

**Comment by tech2serve** [comment, marked as answer]
Here is what I got when I mapped to your meeting:

<img width="794" alt="image" src="https://user-images.githubusercontent.com/31904316/177666948-247b4247-6e97-4253-a2a5-9990479b526e.png">

Is that not correct? :-)

We do generate the URL for the directions, and I will have to spend more time trying to duplicate your issue. This is a bit more difficult that it may sound as there are several factors that could be affecting things. For example, even in my attempt at humor shown above, the starting location isn't correct...I am not in Midtown Manhattan. This is happening because Google doesn't have permission for my actual location, and sometimes geolocation on a laptop doesn't work on that well to begin with.  

So I switched to my iPad (with GPS) and tried mapping to a meeting. This time, Google took an address from my "home". I have no idea why since it isn't my current location (by about 2,000 miles). Forcing Google to update to my true current location did finally work. But, this was because the GPS in my iPad was being used.

All of this is to say that we could look at the code to make sure it isn't in need of update (not sure of the last time it has been touched) but I do suspect that we'd be seeing an error from Google if they'd dropped support for the parameters we use. But, I could be wrong so I'll investigate.

Here is what you could do in the meantime. One thing I note is different on your site than on others I've tried is that you have the dialogue popping up asking for permission to redirect. It is probably not a factor, but if you can disable that and try again? Also, perhaps make sure that if you bring up Google maps in your browser before testing, make sure it can display your current location, and that it is correct. Sometimes I get shown a state or two away!

**Reply by mikepoconnor4** [reply]
Sounds a little pricey if that's the one way price. (rah rah rah)

As you said, there's a few things going on - I think that syntax might still be correct for iOS devices - so you might find situations are different depending on Windows, Android, iOS. 

Yeah I agree with getting rid of that redirect message in the directions case. I'll do that and try again. But I would note that if you use one of the other intergroups, you get the same strange directions - https://aasfmarin.org/find-a-meeting?meeting=larkspur-beginners-592 is an example. 

And also try the link I coded and see if that shows your location properly - it does for me in my setup. https://www.google.com/maps/dir/?api=1&destination=2709+Kalakaua+Ave%2C+Honolulu%2C+HI+96815%2C+USA

**Reply by mikepoconnor4** [reply]
saddr is supported in iOS - https://developers.google.com/maps/documentation/urls/ios-urlscheme

The scheme without saddr is supposed to work for all devices - see https://developers.google.com/maps/documentation/urls/get-started 

They specifcally note 

"Using Maps URLs, you can build a universal, cross-platform URL to launch Google Maps and perform searches, get directions and navigation, and display map views and panoramic images. The URL syntax is the same regardless of the platform in use.

You don't need a Google API key to use Maps URLs.

Universal cross-platform syntax
As a developer of an Android app, an iOS app, or a website, you can construct a common URL, and it will open Google Maps and perform the requested action, no matter the platform in use when the map is opened.

On an Android device:
If [Google Maps app for Android](https://play.google.com/store/apps/details?id=com.google.android.apps.maps) is installed and active, the URL launches Google Maps in the Maps app and performs the requested action.
If the Google Maps app is not installed or is disabled, the URL launches Google Maps in a browser and performs the requested action.
On an iOS device:
If [Google Maps app for iOS](https://itunes.apple.com/app/id585027354) is installed, the URL launches Google Maps in the Maps app and performs the requested action.
If the Google Maps app is not installed, the URL launches Google Maps in a browser and performs the requested action.
On any other device, the URL launches Google Maps in a browser and performs the requested action."

**Reply by mikepoconnor4** [reply]
I got rid of the redirect message - it's still starting from the strange starting point.

**Reply by tech2serve** [reply]
Thank you for trying it without the redirect.

What browser are you using?

Regrettably, I'm probably not the best C4R member to try and duplicate your issue because of the uniqueness of my situation. I'm a nomad, my Internet comes through cell service (and a cellular router) and I'm often in remote locations (like now), so nearly every attempt to get directions is leading to different results. 

To make matters worse, the major browsers use different techniques to determine a user's location, and these techniques also differ from laptop to, say, an iPad. I've seen situations where one browser will work, and another will not. 

In Chrome, I did finally get some directions to a not-too-distant meeting here in New Hampshire:

<img width="209" alt="image" src="https://user-images.githubusercontent.com/31904316/177797562-7e47d714-81c7-434c-a8c4-018b657a6277.png">

I will keep trying though, including (hopefully) creating a TSML branch that will implement the latest Google recommendations.

@code4recovery/tsml-maintainers, @code4recovery/tsml-explorers  I would appreciate if someone else can dive in here and see if we can reliably reproduce Mike's results.

**Reply by tech2serve** [reply]
Ok!  A proof of concept turned out to be easier than I thought. I was able to build the URL using the updated Google guidance and the directions built (shown above) started from my blue location dot.

As I can reproduce this, I'll create an issue and C4R will discuss how to best approach making the change. It will have to occur in a few places.

Thanks for your patience, great research, and bringing it to our attention.

**Reply by mikepoconnor4** [reply]
Great. I saw it fail on Chrome in Windows 11 (i.e. the strange starting location). I noticed different behavior in Chrome on Android - there the map popped up (I think it popped up the Google Maps app) - but the starting location had "Current Location" in it and the destination was just a lat/lon - but more importantly it seemed like Google Maps didn't understand it at all and didn't render a map under it.
